components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/util/LoginContextManagementUtil.java

```diff
@@ -323,14 +325,43 @@ private static boolean isStepHasMultiOption(AuthenticationContext context) {
         return stepHasMultiOption;
     }
 
-    private static boolean isValidRequest(HttpServletRequest request, AuthenticationContext context) {
+    private static boolean canHandleAuthenticator(AuthenticationContext context, String authenticators) {
+
+        List<String> authenticatorsList = new ArrayList();
+        if (authenticators != null) {
+            String[] authenticatorIdPMappings = authenticators.split(";");
+            for (String authenticatorIdPMapping : authenticatorIdPMappings) {
+                String[] authenticatorIdPMapArr = authenticatorIdPMapping.split(":");
+                authenticatorsList.add(authenticatorIdPMapArr[0]);
+            }
+        }
+
+        Map<Integer, StepConfig> stepMap = context.getSequenceConfig().getStepMap();
+        if (stepMap != null && !stepMap.isEmpty()) {
+            StepConfig stepConfig = stepMap.get(context.getCurrentStep());
+
+            for (AuthenticatorConfig authenticatorConfig : stepConfig.getAuthenticatorList()) {
+                if (!authenticatorsList.contains(authenticatorConfig.getName())) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+    private static boolean isValidRequest(HttpServletRequest request, AuthenticationContext context,
+                                          String authenticators) {
 
         boolean isValidRequest = (context != null &&
                 context.getProperty(FrameworkConstants.CURRENT_POST_AUTHENTICATION_HANDLER) == null);
         if (isNonceCookieEnabled() && isValidRequest) {
             Cookie nonceCookie = FrameworkUtils.getCookie(request, getNonceCookieName(context));
             isValidRequest = nonceCookie != null && StringUtils.isNotBlank(nonceCookie.getValue());
         }
+        if (StringUtils.isNotEmpty(authenticators) && isValidRequest) {
```

Comment:
Why are we doing the canHandleAuthenticator check only when isValidRequest is true?

Reply:
Similar to line 358, skipping this extra validation if the sessionDataKey is invalid.